peps/pep-0742.rst

```diff
@@ -0,0 +1,333 @@
+emit an error if this condition is not met.
+
+Formally, type *NP* should be narrowed to :math:`A \land R`, the intersection of *A* and *R*, and type *NN* should be narrowed to
+:math:`A \land \neg R`, the intersection of *A* and the complement of *R*.
+In practice, the theoretic types for strict type guards cannot be expressed
+precisely in the Python type system. Type checkers should fall back on
+practical approximations of these types. As a rule of thumb, a type checker
+should use the same type narrowing logic -- and get results that are consistent
+with -- its handling of :py:func:`isinstance`. This guidance allows for changes and
+improvements if the type system is extended in the future.
+
+Examples
+--------
+
+Type narrowing is applied in both the positive and negative case::
+
+    from typing import TypeNarrower, assert_type
+
+    def is_str(x: object) -> TypeNarrower[str]:
+        return isinstance(x, str)
+
+    def f(x: str | int) -> None:
+        if is_str(x):
+            assert_type(x, str)
+        else:
+            assert_type(x, int)
+
+The final narrowed type may be narrower than R, due to the constraints of the
+argument's previously-known type::
+
+    from collections.abc import Awaitable
+    from typing import Any, TypeNarrower, assert_type
+    import inspect
+
+    def isawaitable(x: object) -> TypeNarrower[Awaitable[Any]]:
+        return inspect.isawaitable(x)
+
+    def f(x: Awaitable[int] | int) -> None:
+        if isawaitable(x):
+            assert_type(x, Awaitable[int])
+        else:
+            assert_type(x, int)
+
+It is an error to narrow to a type that is not consistent with the input type::
+
+    from typing import TypeNarrower
+
+    def is_str(x: int) -> TypeNarrower[str]:  # Type checker error
+        ...
+
+Subtyping
+---------
+
+``TypeNarrower`` is not a subtype of ``bool``.
+The type ``Callable[..., TypeNarrower[int]]`` is not assignable to
+``Callable[..., bool]`` or ``Callable[..., TypeGuard[int]]``, and vice versa.
+This restriction is carried over from :pep:`647`. It may be possible to relax
+it in the future, but that is outside the scope of this PEP.
+
+Unlike ``TypeGuard``, ``TypeNarrower`` is invariant in its argument type:
+``TypeNarrower[B]`` is not a subtype of ``TypeNarrower[A]``, even if ``B`` is a subtype of ``A``.
+To see why, consider the following example::
+
+    def takes_narrower(x: int | str, narrower: Callable[[object], TypeNarrower[int]]):
+        if narrower(x):
+            print(x + 1)  # x is an int
+        else:
+            print("Hello " + x)  # x is a str
+
+    def is_bool(x: object) -> TypeNarrower[bool]:
+        return isinstance(x, bool)
+
+    takes_narrower(1, is_bool)  # Error: is_bool is not a TypeNarrower[int]
+
+(Note that ``bool`` is a subtype of ``int``.)
+This code fails at runtime, because the narrower returns False (1 is not a ``bool``)
```

Comment:
```suggestion
This code fails at runtime, because the narrower returns ``False`` (1 is not a ``bool``)
```
or
```suggestion
This code fails at runtime, because the narrower returns false (1 is not a ``bool``)
```